Railsアプリの初期化

a5ob7r opened the pull request:
# 概要

APIサーバ(https://github.com/sls-training/2023-API-training-server) のクライアント用のRailsアプリを雛形を追加する。

`rails new` で初期化してから、APIサーバ(https://github.com/sls-training/2023-API-training-server) と同様のgemや設定を追加した。

a5ob7r committed:
Add some files for docker compose

a5ob7r committed:
Initialize the Rails application

Initialized as following steps

```sh
bundle init
# Uncomment 'gem "rails"' in Gemfile
bundle install
bundle exec rails new . --database=postgresql --skip-keeps --skip-action-mailer --skip-action-mailbox --skip-action-text --skip-active-job --skip-active-storage --skip-action-cable --skip-asset-pipeline --skip-javascript --skip-hotwire --skip-jbuilder --skip-test --skip-system-test --minimal --skip-bundle
```

a5ob7r committed:
Add Rubocop

a5ob7r committed:
Fix some files in accordance to Rubocop

Fixed some files by the following command and handwriting.

```sh
bin/rubocop -a
```

a5ob7r committed:
Add brakeman

a5ob7r committed:
Add RSpec

Installed by the following command.

```sh
bin/rails g rspec:install
```

a5ob7r committed:
Add rbs collection

Install rbs collection configs by following commands.

```sh
bundle exec rbs collection init
bundle exec rbs collection install
```

a5ob7r committed:
Add rbs_rails

Install by the following command.

```sh
bin/rails g rbs_rails:install
```

a5ob7r committed:
Add katakata_irb

a5ob7r committed:
Update database configs

a5ob7r committed:
Add a database schema

Generated by the following command.

```sh
bin/rails db:migrate
```

a5ob7r committed:
Add a GHA config

a5ob7r committed:
Add a missing binstub for bundler

a5ob7r committed:
Add a missing platform to Gemfile.lock for CI

tk0miya reviewed (approved):
きれいにコミット分かれてて読みやすかったです。

a5ob7r committed:
Add a missing binstub for rspec-daemon

a5ob7r committed:
Change rspec container's port to configurable

Maybe other projects use the same port and if so it causes a port number
conflict. So we make it configurable to map it to another port.

a5ob7r commented:
rspec-daemon用のコンテナのポートを環境変数経由で設定可能にしました。他のプロジェクトでもrspec-daemonを使っている場合に、ポート番号が競合する可能性があるからです。

tk0miya commented on compose.yaml:
```diff
@@ -27,7 +27,7 @@ services:
     command: bin/rspec-daemon
     ports:
-      - 3002:3002
+      - ${RSPEC_PORT:-3002}:3002
```
RSPEC_DAEMON_PORT の方がいいかもなあ

a5ob7r replied:
それは僕も考えたのですが、同様の設定をしている箇所で `コンテナ名_PORT` というフォーマットを利用していたので、そちらに合わせました。

https://github.com/sls-training/2023-API-training-client/blob/e6274a6fbb1074db2d3700068e1bd48248acfb62/compose.yaml#L10-L20

tk0miya replied:
APP_PORT というのはそこまで深く検討して名付けられていないような気がしました。
その命名規約だと、ひとつのコンテナで複数のポートを使う場合に困りそうです。

また、その方向でいくのであればコンテナ名を rspec-daemon としても良さそうです。

上のコメントをしたのは RSPEC_PORT という名前だと、rspec がなにか待ち受けしてそうに見えるので、誤解を生みそうだという気持ち悪さがあったためです。
この手のネーミングは、数年後など、忘れた頃にもやっとしやすいので、正しい命名をしておいたほうがハッピーだと考えています。

※ 流れ上変更しようというアピールに見えますが、短命のプロジェクトなのですごく強い意見ではありません。理解しつつも今回はこれで行こう、という選択はありです。

a5ob7r replied:
> その命名規約だと、ひとつのコンテナで複数のポートを使う場合に困りそうです。

1プロセス1コンテナで運用すると思うのですが、1プロセスで複数のポートを同時利用することってあるんですかね。もちろん可能だとは思いますが。

epaew replied:
Rails に限らなくてよければ、FTP (20-21) とか

a5ob7r replied:
そういえば、HTTP(80)とHTTPS(443)とかもあるか。FTPが利用するポートについては知りませんでした。

epaew replied:
あとは、メール？
- SMTP
- POP3
- IMAP

最近は POP3 なんて見なくなった気はするけど

tk0miya replied:
レアですが、ありますね。データポートとコントロールポートが分かれているようなサーバがあります。たとえば FTP などがいい例ですね (20番と21番を使う)。

ほとんどのケースでは 1つのコンテナにつき 1つのポートなので、この指摘をもとにネーミングを変えるというのは極端だとは思いますが、「コンテナ名_PORT」というルールが絶対だとは考えないほうが良さそうです。

Taku3939 commented on compose.yaml:
```diff
@@ -0,0 +1,47 @@
+
+services:
+  app:
+    <<: *app-template
```
テンプレート作れるの初めて知った。

a5ob7r replied:
エクステンションと言うみたいです。
https://docs.docker.com/compose/compose-file/11-extension/

tim-toga reviewed (approved):
rubocop -aを最後にやった方がより良かったかも

a5ob7r commented:
> rubocop -aを最後にやった方がより良かったかも

@tim-toga そういうものなんでしょうか。

tim-toga commented:
全ての作業を終えて、PRを出す前にrubocopをかけた方が全ての修正やファイルに対して実行できるので、最後に実行した方がよりメリットがあると思います

a5ob7r commented:
あー、そういう運用もありですかね。個人的には、PR単位ではなくコミット単位で常時Rubocopをかけているので、特に最後に実行する必要はないかなと考えています。今回は、Rubocopを追加した後で既存のコードベースに自動修正を適用したので、途中で明示的にコミットが含まれています。

tim-toga commented:
なるほど。それなら良いですね。rubocop -aをするコミットが途中にあったので、作業途中に一回だけ実行したのかと勘違いしました
